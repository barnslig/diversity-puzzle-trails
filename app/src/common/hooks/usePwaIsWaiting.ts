import * as React from "react";

/**
 * A React hook that returns whether the registered SW is waiting
 *
 * @returns Whether the registered SW has been installed but is waiting to activate
 */
const usePwaIsWaiting = () => {
  /**
   * Whether the registered SW is waiting
   */
  const [isWaiting, setIsWaiting] = React.useState(window.swIsWaiting);

  React.useEffect(() => {
    if (!window.wb) {
      return;
    }

    /**
     * A local reference to the workbox instance
     */
    const wb = window.wb;

    /**
     * Event handler for when the SW is waiting
     */
    const onIsWaiting = () => {
      setIsWaiting(true);
    };

    wb.addEventListener("waiting", onIsWaiting);

    return () => {
      wb.removeEventListener("waiting", onIsWaiting);
    };
  }, []);

  return isWaiting;
};

export default usePwaIsWaiting;
