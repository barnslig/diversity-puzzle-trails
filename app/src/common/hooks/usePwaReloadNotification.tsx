import { Button } from "@mui/material";
import { FormattedMessage, useIntl } from "react-intl";
import { useSnackbar } from "notistack";
import * as React from "react";

import usePwaIsWaiting from "./usePwaIsWaiting";

/**
 * A React hook that triggers a notistack notification when a PWA update is available
 */
const usePwaReloadNotification = () => {
  const { enqueueSnackbar } = useSnackbar();
  const intl = useIntl();

  /**
   * Whether the registered SW is waiting to activate
   */
  const isWaiting = usePwaIsWaiting();

  /**
   * Event handler for when the reload button is clicked by the user
   */
  const onClickReload = React.useCallback(() => {
    if (!window.wb) {
      return;
    }

    window.wb.addEventListener("controlling", (event) => {
      window.location.reload();
    });

    window.wb.messageSkipWaiting();
  }, []);

  React.useEffect(() => {
    if (isWaiting) {
      const action = (
        <Button onClick={onClickReload}>
          <FormattedMessage
            defaultMessage="Neu laden"
            description="pwa reload snackbar action button label"
          />
        </Button>
      );

      enqueueSnackbar(
        intl.formatMessage({
          defaultMessage: "Neue Inhalte sind verf√ºgbar.",
          description: "pwa reload snackbar",
        }),
        {
          action,
          key: "pwa-waiting",
          persist: true,
        }
      );
    }
  }, [enqueueSnackbar, intl, isWaiting, onClickReload]);
};

export default usePwaReloadNotification;
