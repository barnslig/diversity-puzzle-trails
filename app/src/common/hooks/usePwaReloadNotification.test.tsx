import { screen } from "@testing-library/react";
import * as React from "react";
import render from "../testing/render";
import type { Workbox, WorkboxEvent } from "workbox-window";

import * as usePwaIsWaiting from "./usePwaIsWaiting";

import usePwaReloadNotification from "./usePwaReloadNotification";
import userEvent from "@testing-library/user-event";

const TestComponent = () => {
  usePwaReloadNotification();
  return null;
};

it("does not show a snackbar when no SW is waiting to be activated", () => {
  jest.spyOn(usePwaIsWaiting, "default").mockReturnValue(false);

  const { container } = render(<TestComponent />);

  expect(
    screen.queryByText(/Neue Inhalte sind verfügbar./)
  ).not.toBeInTheDocument();
  expect(container.firstChild).toMatchSnapshot();
});

it("shows a snackbar with an action when a SW is waiting to be activated", () => {
  jest.spyOn(usePwaIsWaiting, "default").mockReturnValue(true);

  const { container } = render(<TestComponent />);

  expect(screen.getByText(/Neue Inhalte sind verfügbar./)).toBeInTheDocument();
  expect(screen.getByText(/Neu laden/)).toBeInTheDocument();
  expect(container.firstChild).toMatchSnapshot();
});

it("skips waiting and reloads the page when the user clicks the action button", () => {
  jest.spyOn(usePwaIsWaiting, "default").mockReturnValue(true);

  const location = window.location;
  // @ts-ignore
  delete window.location;
  window.location = {
    ...location,
    reload: jest.fn(),
  };

  const wb = new EventTarget() as unknown as Workbox;
  wb.messageSkipWaiting = jest.fn();
  window.wb = wb;

  render(<TestComponent />);

  userEvent.click(screen.getByText(/Neu laden/));

  expect(wb.messageSkipWaiting).toHaveBeenCalled();

  wb.dispatchEvent(
    new Event("controlling") as unknown as WorkboxEvent<"controlling">
  );

  expect(window.location.reload).toHaveBeenCalled();
});
