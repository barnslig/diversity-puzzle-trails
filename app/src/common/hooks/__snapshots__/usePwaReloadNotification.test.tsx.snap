// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`does not show a snackbar when no SW is waiting to be activated 1`] = `null`;

exports[`shows a snackbar with an action when a SW is waiting to be activated 1`] = `
<div
  class="SnackbarContainer-bottom SnackbarContainer-left SnackbarContainer-root css-uwcd5u"
>
  <div
    class="MuiCollapse-root MuiCollapse-vertical MuiCollapse-entered css-pwcg7p-MuiCollapse-root"
    style="min-height: 0px;"
  >
    <div
      class="MuiCollapse-wrapper MuiCollapse-vertical css-smkl36-MuiCollapse-wrapper"
    >
      <div
        class="MuiCollapse-wrapperInner MuiCollapse-vertical css-9l5vo-MuiCollapse-wrapperInner"
      >
        <div
          class="SnackbarItem-wrappedRoot css-96yce9"
        >
          <div
            aria-describedby="notistack-snackbar"
            class="SnackbarContent-root SnackbarItem-contentRoot css-hped4j"
            role="alert"
            style="webkit-transform: none; transform: none; webkit-transition: -webkit-transform 225ms cubic-bezier(0.0, 0, 0.2, 1) 0ms; transition: transform 225ms cubic-bezier(0.0, 0, 0.2, 1) 0ms;"
          >
            <div
              class="SnackbarItem-message"
              id="notistack-snackbar"
            >
              Neue Inhalte sind verf√ºgbar.
            </div>
            <div
              class="SnackbarItem-action"
            >
              <button
                class="MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButtonBase-root  css-1e6y48t-MuiButtonBase-root-MuiButton-root"
                tabindex="0"
                type="button"
              >
                Neu laden
                <span
                  class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                />
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
